libarchive/test/test_write_disk_secure_noabsolutepaths.c

```diff
@@ -0,0 +1,102 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2024 Mostyn Bramley-Moore <mostyn@antipode.se>
+ */
+
+#include "test.h"
+
+#if defined(_WIN32) && !defined(__CYGWIN__)
+
+#include <fileapi.h>
+
+#define UNLINK _unlink
+
+#else
+
+#include <stdlib.h>
+#include <unistd.h>
+
+#define UNLINK unlink
+
+#endif
+
+/*
+ * Exercise security checks that should prevent writing absolute paths
+ * when extracting archives.
+ */
+DEFINE_TEST(test_write_disk_secure_noabsolutepaths)
+{
+	struct archive *a, *ad;
+	struct archive_entry *ae;
+
+	char buff[10000];
+
+	size_t used;
+
+	// Create an archive_write object.
+	assert((a = archive_write_new()) != NULL);
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_format_ustar(a));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_add_filter_none(a));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_open_memory(a, buff, sizeof(buff), &used));
+
+#if defined(_WIN32) && !defined(__CYGWIN__)
+	char temp_dir[MAX_PATH + 1];
+	assert(GetTempPathA(MAX_PATH + 1, temp_dir) != 0);
+
+	char temp_file_name[MAX_PATH + 1];
+	char temp_absolute_file_name[MAX_PATH + 1];
+	assert(GetTempFileNameA(temp_dir, "abs", 0, temp_file_name) != 0);
```

Comment:
As above, using `testworkdir` here instead of `temp_dir` would help ensure that we never pollute the temp directory more than we have to.